Chepa kn unit5

ChepaKN opened the pull request.

ChepaKN committed:
getOptimalTransport() realization. added Main.java

ChepaKN committed:
поправил реализацию getOptimalTransport
в процессе реализации inversePathFinder
в BigDecimalRange добавил метод inRange

ChepaKN committed:
realization of getOptimalTransport and getAllSolutions methods.
method inRange are added in BigDecimalRange class;
Надо бы приделать использование Otpional для защиты от NPE.

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/PathFinder.java:
```diff
@@ -14,6 +18,36 @@
         // ToDo: realize me!
-        return null;
+
+        if (deliveryTask == null) return null;
```
if, даже однострочный - всегда обрамляем фигурными скобками и тело с новой строки. Так удобнее читать код.

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/PathFinder.java:
```diff
@@ -14,6 +18,36 @@
+
+        //Упакуем в map, чтобы удабно было узнать длину маршрута для заданного типа по ключу RouteType
+        Map<RouteType, BigDecimal> routesMap = new HashMap<>();
+        for (Route r : deliveryTask.getRoutes()) {
```
Лекция была по ФП :-)
И, хотя явно сказано не было, но не рекомендуется использовать циклы. Цель - попрактиковаться в стримах.

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/PathFinder.java:
```diff
@@ -14,6 +18,36 @@
+                .sorted(Comparator.comparing(o -> o.getPrice().multiply(routesMap.get(o.getType()))))
+                .collect(Collectors.toList());
+
+        return  sortedTransport.get(0);
```
А если List - пустой?

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/inverse/BigDecimalRange.java:
```diff
@@ -206,4 +206,34 @@ public static BigDecimalRange leftStrictRightOpenRange(BigDecimal left) {
         return new BigDecimalRange(null, true, null, true);
     }
+
+    public boolean inRange(BigDecimal value){
```
Здесь все отлично. И хорошо, что этот метод внутри класса BigDecimalRange.

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/InversePathFinder.java:
```diff
@@ -25,6 +31,78 @@
+        //Диапазон затрат
+        BigDecimalRange range = task.getPriceRange();
+
+        //todo: как бы тут прикрутить стрим...
```
Посмотрите в сторону flatMap / reduce.
Хотя их применение "зависит", конечно, от того как именно Вы будете решать задачу.

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/InversePathFinder.java:
```diff
@@ -25,6 +31,78 @@
+
+            //Добавим в коллекцию все решения
+            if (availableTransport.size() > 0) {
+                for (Transport t : availableTransport) {
```
И снова цикл. Попробуйте избавиться :-)

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/InversePathFinder.java:
```diff
@@ -25,6 +31,78 @@
+        }
+
+        //Сортируем список
+        Comparator<Solution> solutionComparator = Comparator.comparing(Solution::getPrice).reversed().
```
Сортировка +

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/InversePathFinder.java:
```diff
@@ -25,6 +31,78 @@
+    private Map<RouteType, BigDecimal> routesToMap(DeliveryTask deliveryTask){
+            //Упакуем в map, чтобы удабно было узнать длину маршрута для заданного типа по ключу RouteType
+            Map<RouteType, BigDecimal> routesMap = new HashMap<>();
+            for (Route r : deliveryTask.getRoutes()) {
```
reduce / flatMap

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/Main.java:
```diff
@@ -0,0 +1,47 @@
+    public static final Transport Tanker = new Transport("Танкер", valueOf(200), valueOf(100), RouteType.Sea);
+    public static final Transport Helicopter = new Transport("Вертолет", valueOf(200), valueOf(100), RouteType.Air);
+
+    public static void main(String[] args){
```
В принципе - для обеих задач есть тесты. Можно (и даже нужно) проверять через них. Или Вы просто для себя еще пример набросали?

pavelsilent replied:
Хотя я смотрю в итоговом мердже вроде этих чейнджей нет.

pavelsilent commented on unit5/unit5-homework/src/main/java/pro/sisit/javacourse/InversePathFinder.java:
```diff
@@ -25,6 +31,78 @@
-        return new ArrayList<>();
+
+        List<Solution> solutions = new ArrayList<>();
+        if ((task == null) || (task.getTasks() == null) || (task.getTransports() == null) || (task.getPriceRange() == null)){
```
if'ы можно попробовать заменить на Optional.

pavelsilent commented:
Посмотрел задачу. В целом все правильно, но местами не совсем по "концепции" ФП.
Из доработок - нужно избавиться от всех циклов. 
Из улучшений - попробуйте воспользоваться Optional - увидите они сильно упрощают жизнь (а еще их можно комбинировать со стримами).

Также хочу обратить внимание, что когда будете переделывать на стримы - не используйте внутри лямбд списки / хэши, объявленные снаружи. Java это позволит, но таким образом мы всеравно будет мутировать эту переменную, что есть не очень хорошо.

ChepaKN committed:
Refactoring: Постарался переделать некоторые моменты в программе ближе к концепции ФП. Было несколько замечаний по наличию циклов -> исправлено.
По поводу Optional: Немного непонятно, чем лучше проверка if(!Optional.ofNullable(task).isPresent()) чем if(task == null)...

pavelsilent commented:
Добрый день. Планируете еще что-то доделывать по этому заданию?

ChepaKN commented:
Да, прошу прощения за задержку, optional мне не зашёл, отвлекся на unit7.
Конечно хочу переделать

сб, 4 апр. 2020 г., 16:00 Pavel Brichev <notifications@github.com>:

> Добрый день. Планируете еще что-то доделывать по этому заданию?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/ChepaKN/JavaSIS-3.20/pull/8#issuecomment-608998859>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKWSJO6YU4MDHTX3KYA4TF3RK3ZKPANCNFSM4LSBOWDQ>
> .
>

pavelsilent commented:
Ок, отлично. Ничего страшного, я просто мониторил решения и решил уточнить.
